telemetry/tracks/class-tracks-event.php

```diff
@@ -0,0 +1,336 @@
+<?php
+/**
+ * Telemetry: Tracks Event class
+ *
+ * @package Automattic\VIP\Telemetry\Tracks
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry\Tracks;
+
+use WP_Error;
+use JsonSerializable;
+use Automattic\VIP\Support_User\User as Support_User;
+use function Automattic\VIP\Logstash\log2logstash;
+
+/**
+ * Class that creates and validates Tracks events.
+ *
+ * @since 3.12.0
+ */
+class Tracks_Event implements JsonSerializable {
+	/**
+	 * Event name regex. Spaces, mixed case, and special characters are not allowed.
+	 */
+	protected const EVENT_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Property name regex. Event props should be in snake_case. Example: compressed_size is correct, but compressedSize is not.
+	 * Property names with leading underscores are reserved for special properties.
+	 */
+	protected const PROPERTY_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Variable containing the event's data or an error if one was encountered
+	 * during the event's creation.
+	 *
+	 * @var Tracks_Event_DTO|WP_Error
+	 */
+	protected $data = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string                            $prefix The event's prefix.
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any properties included in the event.
+	 */
+	public function __construct( string $prefix, string $event_name, array $event_properties = [] ) {
+		$event_data        = static::process_properties( $prefix, $event_name, $event_properties );
+		$validation_result = static::get_event_validation_result( $event_data );
```

Comment:
There's very little benefit to static methods here (they don't get to access `$this` and thus somewhat faster), so if there are strong feelings against using them, we can switch to instance methods instead.